core/src/term/pattern/bindings.rs

```diff
@@ -0,0 +1,181 @@
+//! Pattern analysis.
+
+use crate::{
+    identifier::LocIdent,
+    term::{pattern::*, record::Field},
+};
+
+pub trait Bindings {
+    /// Returns a list of all variables bound by this pattern, together with the path to the field
+    /// they match and the associated extra annotations.
+    ///
+    /// # Example
+    ///
+    /// For a pattern `{a = x @ {foo, bar | Number = z, ..rest}, d = e}`, the result of this function
+    /// contains:
+    ///
+    /// - `(["a"], "x", empty field)` for the `x` variable
+    /// - `(["a"], "rest", empty field)` for the `rest` variable
+    /// - `(["a", "foo"], "foo", empty field)` for the `foo` variable
+    /// - `(["a", "bar"], "z", field with Number contract)` for the `z` variable
+    /// - `(["d"], "e", empty field)` for the `e` variable
+    fn bindings(&self) -> Vec<(Vec<LocIdent>, LocIdent, Field)>;
```

Comment:
Was the code of this module taken from the LSP or something? I feel like I've seen this code somewhere else before. If yes, has it been removed from the LSP?But maybe it's memories from my previous review :man_shrugging: